File: tests/test_config_syntax.mjs
#!/usr/bin/env node
/**
 * 契约测试：wails.json + go.mod 语法与结构校验（reasonix.toml 为本地 AI 终端配置，不入库不校验）。
 * 由 tests/python/test_config_syntax.py 迁移（2026-08-03），校验逻辑逐点保真。
 */
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');

function checkWails() {
  const errors = [];
  const fp = path.join(ROOT, 'wails.json');
  if (!fs.existsSync(fp)) {
    errors.push('MISSING: wails.json');
    return errors;
  }

  let data;
  try {
    data = JSON.parse(fs.readFileSync(fp, 'utf-8'));
  } catch (e) {
    errors.push(`SYNTAX: wails.json 解析失败: ${e.message}`);
    return errors;
  }

  // Wails 3 契约：v2 平铺结构（outputfilename / frontend:* / bind）已弃用。
  // 迁移依据见 docs/architecture/adr/ADR-001-wails3-migration.md §3。
  if (!data.name) {
    errors.push("'name' must be non-empty");
  }

  const schema = data.$schema ?? '';
  if (!schema.includes('v3.wails.io')) {
    errors.push("'$schema' 必须指向 v3.wails.io（v2 配置不应残留）");
  }

  const frontend = data.frontend;
  if (typeof frontend !== 'object' || frontend === null) {
    errors.push("'frontend' 必须是对象");
  } else {
    if (!frontend.dir) errors.push("'frontend.dir' must be non-empty");
    if (!frontend.install) errors.push("'frontend.install' must be non-empty");
    if (!frontend.build) errors.push("'frontend.build' must be non-empty");
  }

  // v2 残留守卫：v3 已弃用顶层 bind 字段（service 自动发现替代）
  if ('bind' in data) {
    errors.push("'bind' 字段在 v3 已弃用（v2 残留，应移除）");
  }
  return errors;
}

function checkGomod() {
  const errors = [];
  const fp = path.join(ROOT, 'go.mod');
  if (!fs.existsSync(fp)) {
    errors.push('MISSING: go.mod');
    return errors;
  }

  const text = fs.readFileSync(fp, 'utf-8').replace(/\r\n/g, '\n');
  const lines = text.split('\n');

  if (!text.startsWith('module ')) {
    errors.push("must start with 'module <name>'");
  }

  let goVersion = null;
  for (const line of lines) {
    const m = line.match(/^go\s+(\S+)/);
    if (m) {
      goVersion = m[1];
      break;
    }
  }
  if (!goVersion) {
    errors.push("missing 'go X.Y.Z' version line");
  } else {
    const parts = goVersion.split('.');
    if (parts.length >= 2) {
      const major = parseInt(parts[0], 10);
      const minor = parseInt(parts[1], 10);
      if (Number.isNaN(major) || Number.isNaN(minor)) {
        errors.push(`invalid go version '${goVersion}'`);
      } else if (major < 1 || minor < 20) {
        errors.push(`go version ${goVersion} too old, need 1.20+`);
      }
    }
  }

  let requireStart = -1;
  let requireEnd = -1;
  for (let i = 0; i < lines.length; i++) {
    if (/^require\s*\($/.test(lines[i])) requireStart = i;
    if (requireStart >= 0 && lines[i].trim() === ')') {
      requireEnd = i;
      break;
    }
  }
  if (requireStart < 0) {
    errors.push("missing 'require (...)' block");
  } else if (requireEnd - requireStart < 2) {
    errors.push('too few dependencies in require block');
  }

  return errors;
}

const errors = [];
for (const e of checkWails()) errors.push(['wails.json', e]);
for (const e of checkGomod()) errors.push(['go.mod', e]);

if (errors.length) {
  console.error(`FAILED: ${errors.length} issue(s)\n`);
  for (const [src, e] of errors) console.error(`  [${src}] ${e}`);
  process.exit(1);
} else {
  console.log('OK: wails.json + go.mod syntax checks passed');
}

File: tests/python/test_config_syntax.py
#!/usr/bin/env python3
"""契约测试：wails.json + go.mod + reasonix.toml 语法与结构校验。"""
import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent.parent


def check_wails():
    errors = []
    fp = ROOT / "wails.json"
    if not fp.exists():
        errors.append("MISSING: wails.json")
        return errors

    try:
        data = json.loads(fp.read_text("utf-8"))
    except json.JSONDecodeError as e:
        errors.append(f"SYNTAX: wails.json 解析失败: {e}")
        return errors

    # Wails 3 契约：v2 平铺结构（outputfilename / frontend:* / bind）已弃用。
    # 迁移依据见 docs/architecture/adr/ADR-0001-wails3-migration.md §3。
    if not data.get("name"):
        errors.append("'name' must be non-empty")

    schema = data.get("$schema", "")
    if "v3.wails.io" not in schema:
        errors.append("'$schema' 必须指向 v3.wails.io（v2 配置不应残留）")

    frontend = data.get("frontend")
    if not isinstance(frontend, dict):
        errors.append("'frontend' 必须是对象")
    else:
        if not frontend.get("dir"):
            errors.append("'frontend.dir' must be non-empty")
        if not frontend.get("install"):
            errors.append("'frontend.install' must be non-empty")
        if not frontend.get("build"):
            errors.append("'frontend.build' must be non-empty")

    # v2 残留守卫：v3 已弃用顶层 bind 字段（service 自动发现替代）
    if "bind" in data:
        errors.append("'bind' 字段在 v3 已弃用（v2 残留，应移除）")
    return errors


def check_gomod():
    errors = []
    fp = ROOT / "go.mod"
    if not fp.exists():
        errors.append("MISSING: go.mod")
        return errors

    text = fp.read_text("utf-8", errors="replace")
    lines = text.split("\n")

    if not text.startswith("module "):
        errors.append("must start with 'module <name>'")

    go_version = None
    for line in lines:
        m = re.match(r"^go\s+(\S+)", line)
        if m:
            go_version = m.group(1)
            break
    if not go_version:
        errors.append("missing 'go X.Y.Z' version line")
    else:
        parts = go_version.split(".")
        if len(parts) >= 2:
            try:
                if int(parts[0]) < 1 or int(parts[1]) < 20:
                    errors.append(f"go version {go_version} too old, need 1.20+")
            except ValueError:
                errors.append(f"invalid go version '{go_version}'")

    require_start = require_end = -1
    for i, line in enumerate(lines):
        if re.match(r"^require\s*\($", line):
            require_start = i
        if require_start >= 0 and line.strip() == ")":
            require_end = i
            break
    if require_start < 0:
        errors.append("missing 'require (...)' block")
    elif require_end - require_start < 2:
        errors.append("too few dependencies in require block")

    return errors


def check_reasonix():
    errors = []
    fp = ROOT / "reasonix.toml"
    if not fp.exists():
        errors.append("MISSING: reasonix.toml")
        return errors

    text = fp.read_text("utf-8", errors="replace")

    if "config_version" not in text:
        errors.append("missing 'config_version'")
    if "default_model" not in text:
        errors.append("missing 'default_model'")
    if "[agent]" not in text:
        errors.append("missing [agent] section")
    if "[permissions]" not in text:
        errors.append("missing [permissions] section")
    if "[sandbox]" not in text:
        errors.append("missing [sandbox] section")
    return errors


def main():
    errors = []
    errors += [("wails.json", e) for e in check_wails()]
    errors += [("go.mod", e) for e in check_gomod()]
    errors += [("reasonix.toml", e) for e in check_reasonix()]

    if errors:
        sys.stdout.buffer.write(f"FAILED: {len(errors)} issue(s)\n\n".encode("utf-8"))
        for src, e in errors:
            sys.stdout.buffer.write(f"  [{src}] {e}\n".encode("utf-8"))
        sys.exit(1)
    else:
        sys.stdout.buffer.write(b"OK: wails.json + go.mod + reasonix.toml syntax checks passed\n")


if __name__ == "__main__":
    main()
